### PR TITLE
Virtual AES Driver

### PR DESCRIPTION
**Pull request by Ouyancheng**
### Pull Request Overview

This pull request adds the virtualizer for the AES driver 

### Testing Strategy

This pull request was tested by
- File boards/imix/src/test/virtual_aes_ccm_test.rs contains testing with 3 clients. Each of the client is performing the tests specified at capsules/src/test/aes_ccm.rs 

### TODO or Help Wanted

- Currently we couldn't remove the original aes_ccm.rs file, as it's used by somewhere else. There should be further transitions to only use the virtual_aes_ccm driver. 

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

File: capsules/src/virtual_aes_ccm.rs
Comment (niklasad1) [inline capsules/src/virtual_aes_ccm.rs:165]: would be nicer to just pass/borrow `CryptParameters` instead of each of its entries AFAIU.

File: capsules/src/virtual_aes_ccm.rs
Comment (niklasad1) [inline capsules/src/virtual_aes_ccm.rs:100]: please explicit variable names, `a_off` is hard to understand I think `authdata_offset` is better or something similar.

File: capsules/src/virtual_aes_ccm.rs
Comment (niklasad1) [inline capsules/src/virtual_aes_ccm.rs:177]: Do we want to indicate error/debug log if `buf` or `client` is None?

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:100]: Ah sorry, it was originally called a_off in aes_ccm.rs, I don't know what they exactly mean TBH... sorry...

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:178]: added checks for buf / client being None in recent commit.

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs]: Yeah you are right! Thanks for pointing out!

Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs]: resolved in recent commit~

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:100]: So do you want me to expand those names? I am not very sure about them and there's a risk of mis-interpretation... Sorry for this... The code is from aes_ccm.rs.

Comment (niklasad1) [reply capsules/src/virtual_aes_ccm.rs:100]: Let's hear what the others think but if no one understands what they mean that's a problem :)

Could be maybe fixed in another PR then.

Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:100]: Thanks Niklas!

File: capsules/src/virtual_aes_ccm.rs
Comment (bradjc) [inline capsules/src/virtual_aes_ccm.rs:183]: Unfortunately it is problematic to have an error callback here. The issue is that when the user calls `crypt()` and nothing else is using AES, do_next_op runs immediately. If there is an error, this callback is called synchronously. However, tock requires that callbacks _only_ be generated stemming from interrupts.

The general way to solve this is to use a deferred call in `crypt()` so that do_next_op is only ever called from a callback context. The virtual_uart driver illustrates this.

**Review by bradjc — COMMENTED** [review]
I think this should either be merged with the existing aes_ccm capsule, or replace it. The virtualizer layer shouldn't have to duplicate all of the logic from the aes_ccm capsule.

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:216]: Thanks Brad, I've now modified it to use dynamic deferred call to call `do_next_op()` in `crypt()`

Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:216]: > I think this should either be merged with the existing aes_ccm capsule, or replace it. The virtualizer layer shouldn't have to duplicate all of the logic from the aes_ccm capsule.

Yeah, there should be further transitions to use the virtualized driver. But currently I couldn't perform such transition as aes_ccm.rs is used somewhere else. Probably we should do the transition later. Anyways, the virtual_aes_ccm is also functional.

**Comment by hudson-ayers** [comment]
I pushed a commit that changes all uses of the old non-virtualized AESCCM to use the new, virtualized one.

**Comment by Ouyancheng** [comment]
> I pushed a commit that changes all uses of the old non-virtualized AESCCM to use the new, virtualized one.

Thanks a lot Hudson!!!!!!

**Review by hudson-ayers — CHANGES_REQUESTED** [review]
Mostly small changes

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs]: debug stuffs are removed in recent commit.

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs]: Yep, removed in recent commit.

**Review by hudson-ayers — APPROVED** [review]
This PR has the same issue as several other virtualizers that certain types of errors are replaced with debug statements, because there is complexity required to ensure those errors are reported via callbacks, and that those callbacks are only called in callback context. I am not sure it is fair to block this PR on that issue when it exists in many of the other virtual capsules.

**Comment by hudson-ayers** [comment]
This does need to be rebased on master before it can be merged

**Comment by hudson-ayers** [comment]
Rebased/squashed onto master (#2069 caused several conflicts)

**Comment by hudson-ayers** [comment]
I would appreciate you running the AES hardware tests again to confirm I didn't break anything

**Comment by Ouyancheng** [comment]
> I would appreciate you running the AES hardware tests again to confirm I didn't break anything

virtual_aes_ccm_test.rs tests passed!

**Review by hudson-ayers — APPROVED** [review]

File: boards/imix/src/test/virtual_aes_ccm_test.rs
Comment (alistair23) [inline boards/imix/src/test/virtual_aes_ccm_test.rs:33]: Can we remove this commented out code?

File: boards/imix/src/test/virtual_aes_ccm_test.rs
Comment (alistair23) [inline boards/imix/src/test/virtual_aes_ccm_test.rs:2]: Can this be ` ```rust ` instead?

File: boards/imix/src/test/virtual_aes_ccm_test.rs
Comment (alistair23) [inline boards/imix/src/test/virtual_aes_ccm_test.rs:104]: Can you put this note in the header instead?

File: capsules/src/virtual_aes_ccm.rs
Comment (alistair23) [inline capsules/src/virtual_aes_ccm.rs:148]: Why is this commented out?

Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:148]: Hi Alistair, 
The device should be globally enabled in the board's related file instead of here. It's same semantic for the original AES CCM driver. 
Regards, 
Yancheng Ou

File: boards/imix/src/test/virtual_aes_ccm_test.rs
Comment (Ouyancheng) [reply boards/imix/src/test/virtual_aes_ccm_test.rs]: resolved in recent commit

File: boards/imix/src/test/virtual_aes_ccm_test.rs
Comment (Ouyancheng) [reply boards/imix/src/test/virtual_aes_ccm_test.rs]: resolved in recent commit

File: boards/imix/src/test/virtual_aes_ccm_test.rs
Comment (Ouyancheng) [reply boards/imix/src/test/virtual_aes_ccm_test.rs]: resolved in recent commit

File: capsules/src/virtual_aes_ccm.rs
Comment (hudson-ayers) [reply capsules/src/virtual_aes_ccm.rs:148]: I think this line should either be deleted or commented in. I think it makes the most sense to just enable it here, even if the original driver did not do it that way. That helps prevent board authors from forgetting the enable call

**Review by bradjc — APPROVED** [review]

File: capsules/src/virtual_aes_ccm.rs
Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs:148]: OK~ then let me make another commit and uncomment the line~ 
Sorry for this... 

Regards, 
Yancheng Ou

Comment (Ouyancheng) [reply capsules/src/virtual_aes_ccm.rs]: It's uncommented in recent commit. And the tests are passed. 

Regards, 
Yancheng Ou

**Review by alistair23 — APPROVED** [review]

**Review by bradjc — APPROVED** [review]

**Comment by hudson-ayers** [comment]
I am very confused as to how that single line change broke all those tests in a seemingly unrelated area of code

**Comment by Ouyancheng** [comment]
> I am very confused as to how that single line change broke all those tests in a seemingly unrelated area of code

I guess it's because Nano33BLE uncommented its wireless component, which is not fully compatible with the virtualized code.

**Comment by hudson-ayers** [comment]
Ah, okay. I'll rebase.

**Comment by Ouyancheng** [comment]
> Ah, okay. I'll rebase.

Thanks Hudson!

**Review by hudson-ayers — APPROVED** [review]

**Pull request merged**
